### DISCUSSION TITLE
Let's devise a brand new, optimized keymap

### DISCUSSION THREAD

**Discussion by I-am-Erk** [CleverRaven/Cataclysm-DDA#59678] [Q&A]
See https://github.com/CleverRaven/Cataclysm-DDA/issues/59675 for context. This discussion assumes selectable keybinding sets, so that fans of the current layout can quickly switch back. In that context, I think it's high time we rebuild a keybinding map from scratch with no concern for what used to be bound to what key. This isn't something one person can reasonably do, as it should service multiple different play styles across several mods. 

This is *not* a UI discussion, this discussion will assume that we're not moving anything to a new menu or reorganizing windows.

To start, we should compile a list of functions that really must be accessible from a single key / shift-key / ctrl-key press.

---

### Recommendations so far
- movement (10directional)
- player menu
- inventory
- crafting menu
- construction menu
- overmap
- interact
- look around
- smash
- get
- grab
- directional drop
- drag items on ground
- control vehicle
- missions menu
- shout/talk
- movement mode menu
- fighting style menu
- setup zones menu
- zone activities menu

### Up for debate
- toggle run
- toggle crouch
- sleep/wait: should these be inside a context menu with both options?
- Save and quit: Does this need to be top level or just kept in the ESC menu

---

Once this list is done we can use the space below to group these commands into logical sets that should be physically close to each other, such as on the same key or adjacent keys, and also attempt to group them into "things you need to press repeatedly and often" and "not that".

(fill this space later)

---

Finally we can use a keymap image like this and fill out the role of each key.

![image](https://user-images.githubusercontent.com/45136638/181390074-efa59797-d40e-40a0-8a89-f867037902b2.png)

---

#### Numpad
A default layout should make huge use of the numpad. This would require a little (or a lot) more infrastructure so that the keycodes for numpad directional input are understood, instead of requiring numlock to be on to use the numpad. Once we have this we could have a binding like:

- numpad directional: Walk in a direction
    - Ctrl+dir: Sneak in a direction.
    - Shift+dir: Run in a direction
    - Alt+dir: Look in a direction

- Unmodified numpad center button: Open movement mode menu
    - Ctrl+center: Toggle sneak mode on
    - Shift+center: Toggle run mode on
    - Alt+center: center view on player

- Numpad `/`: Drag items on ground
    - Ctrl+`/`: Armour layering
    - Shift+`/`: Open inventory
    - Alt+`/`: Open AIM
- Numpad `*`: Treat as up directional arrow
    - Ctrl+`*`: Sneak up
    - Shift+`*`: Run up
    - Alt+`*`: Peek up
- Numpad `-`: Down directional arrow
    - Ctrl+`-`: Sneak down
    - Shift+`-`: Run down
    - Alt+`-`: Climb down
- Numpad `+`: Interact key
    - Ctrl+`+`: Directional drop
    - Shift+`+`: Get
    - Alt+`+`: Grab
- Numpad enter: Context menu
    - Ctrl+`enter`: Talk/shout/call
    - Shift+`enter`: Control vehicle
    - Alt+`enter`: Grab
- Numpad 0: Fire weapon
    - Ctrl+`0`: Throw item
    - Shift+`0`: Smash
    - Alt+`0`: reload
- Numpad del: Fighting style
    - Ctrl+`del`: Bionics control menu
    - Shift+`del`: Mutations control menu
    - Alt+`del`: do zone activities

In this binding set we'd expect the left hand to cluster on the ctrl-alt-shift keys, and so would have quick access to the left side of the keyboard for the things needed slightly less often:

- q: Quicksave
    - Shift+`q`: Save and quit
- w: Wield
    - Shift+`w`: Wear
    - Ctrl+`w`: Take off
    - Alt+`w`: Armour layering menu
- e: Interact
    - Shift+`e`: Eat
    - Ctrl+`e`: Read
- r: Reload
    - Shift+`r`: Unload
- a: Activate item
    - Shift+`a`: Activate wielded item
    - Ctrl+`a`: Item action menu
    - Alt+`a`: Read
- s:
    - Shift+`s`: 
    - Ctrl+`s`: Toggle safe mode
    - Alt+`s`: Ignore enemy
- d:
- f: fire weapon
    - Shift+`f`: Firing mode
    - Ctrl+`f`: Throw item
    - Alt+`f`: Smash
- c: Close door
    - Shift+`c`: 
    - Ctrl+`c`: 
    - Alt+`c`: Open door
- v: view menu (view nearby items/monsters)
    - Shift+`v`: Look around
    - Ctrl+`v`: Peek
- z: zoom in
    - Shift+`z`: zoom out
    - Ctrl+`z`: Zone management menu
    - Alt+`z`: Do zone activities
- spacebar: pass a turn
    - Shift+`space`: Wait
    - Ctrl+`space`: Sleep
    - Alt+`s`: Ignore enemy
- tab: auto-attack
- 1: Inventory
    - Ctrl+`1`: AIM
- 2: Character sheet
    - Shift+`1`: Diary
- 3: Missions menu
    - Shift+`3`: Factions menu
- 4: Craft
    - Shift+`4`: Construction
    - Ctrl+`4`: Repeat last craft
- 5: Overmap

**Comment by ghost** [comment]
- Imo, having B, N, H, J, K, L, Y, and U all be dedicated to movement is a waste. I know it's old school Roguelike controls and this could be controversial..

- Spacebar can be utilized for more than ESC for menus. During normal gameplay it can open INV, or when the new, multi-tabbed Character/all-in-one menu comes someday, it can open that.

- Also, i love L/R tilt options and think they could be default - SHIFT + L/R arrow keys navigating upwards diagonally and downwards diagonally with CTRL + L/R arrow keys. Chances are slim those combinations will be used for anything else, and if ppl prefer a different modifier with arrow keys we have different options.

**Reply by I-am-Erk** [reply]
BNHJKLYU movement should be in a separate keymap but not part of default, I agree.

We do have several context relevant menus for commands, such as ENTER currently. Space could use one of those, or take over a new one

While I love L/R tilt I don't know if it needs to be default, though that would call attention to it as a nice feature.

**Reply by ghost** [reply]
![CDDA-keybindings](https://user-images.githubusercontent.com/92204064/181392177-b8653f21-8a0c-440b-8455-dfec46688899.jpg)

Stolen Keybindings image from reddit for reference as well (it's outdated but still useful)

from https://www.reddit.com/r/cataclysmdda/comments/ftlpxi/cdda_controls_cheat_sheet/

**Reply by ParaplegicRacehorse** [reply]
As a player who owns zero keyboards with numpad, I disagree that y,u,h,j,k,l,b,n movement keys should be removed from default keymap.

While I may be in a minority, here, I feel really strongly about this.

**Reply by I-am-Erk** [reply]
You're missing the point. You have non-standard keyboards, so there should be a keyset *specifically for* your keyboard. You shouldn't be using the default keyset, because you're not using a default keyboard. And the default keyset shouldn't be designed for your miniature keyboards.

**Reply by Zireael07** [reply]
Majority of laptops do not have a numpad. Doesn't matter whether it's 13" or 17". I wouldn't call a laptop keyboard a "nonstandard one" - I believe the default keyset should take that into account

**Reply by I-am-Erk** [reply]
Absolutely not, that's the entire point. There should be a laptop layout optimized for use without direction pads. The whole gist of this is that with predefined keysets, we don't have to fill the most common letters of the alphabet with a bunch of duplicate functions.

**Reply by BlueKanary** [reply]
As someone currently without a numpad, I think the numpad bindings should be default. It's just so much more intuitive and opens up so much space for less contrived button mapping, like the movement mode menu or stacking morale and item searching. Though a non-numpad keybind set should be relatively easy to access.

**Reply by linonetwo** [reply]
I agree, HJKL are for old keyboards that don't have arrow keys. Player that don't have num pad or don't want to use arrow keys can use a special traditional key set instead. And modern keyboard have at least arrow keys, using them will free lots of keys for binding. For Generation Z and later, we are used to the arrow keys.

**Reply by ShenMian** [reply]
@linonetwo The vi key layout is not just for keyboards without arrow keys. This ergonomic key layout allows users to control the character's movement more efficiently, which the arrow keys cannot do and the arrow keys can only move in a cross shape.

**Comment by Zireael07** [comment]
Does that mean we solved the problem of iirc curses not working with shift/ctrl keybinds which limited alternative functions?

Same question for F1-F12 row which could be useful for stuff like help, factions, achievements, diary etc.

**Comment by chrispikula** [comment]
Why have the look keys be their own set?  Why not have a sub menu to modify your look based on just one?  Press something like Shift-L, to open up the menu, move your center view, (perhaps even with the cursor, gasp!), and done.

Currently, if I ever want to look around, I'm using ';'.  If I ever want to press one of the move camera keys, it's to fix an accidentally keypress once I notice my characters view has been off-centered.

In addition, yeah, using the CTRL keys to do things is great!

I've been using ctrl-D for diary, myself. The ctrl-keys are much nicer than the shift-number keys. Ctrl-V, vehicle, Ctrl-C, craft, Ctrl-X, construct, etc, ctrl-S, Sky(weather), ctrl-F, Forage options, ctrl-A, all the item use options such as star a fire & cut bars, etc.

Lastly, one thing to consider when re-doing the keybindings, we'll likely want to avoid making them entirely use English mnemonics, yeah?  I know all of my example are that, but they are anecdotal, not universal.

**Reply by I-am-Erk** [reply]
I'd suggest that we make ctrl+directional keypad shift your look around, with ctrl+center recentering it, personally.

**Comment by x-qq** [comment]
* There should be at least two default control schemes: for keyboards with a numpad and without
* Obviously the control scheme for keyboards with a numpad should not have roguekeys
* The game should recognize numpad number keys as distinct from the primary number keys (currently it does not, at least on linux-tiles)
* Use of compound keyboard shortcuts should be very minimal (preferably none at all); in particular, often used commands like peek, character screen, craft, build should not require pressing shift to access (currently shift-x, shift-2, shift-7, shift-8)
* Some proposed keys:
  * q - character screen
  * b - build
  * c - craft
  * k - close
  * space - inventory (but available with the i key as well)
  * x - peek (look mode is only ;)

**Reply by I-am-Erk** [reply]
There will definitely need to be a no-keypad layout, but I'm thinking that first we should design the "default" keyboard layout, for the standard 105-key us keyboard.

**Reply by ghost** [reply]
I actually like the idea of TAB becoming the character screen. Especially if/when we combine and make a newer main menu with a multitabbed character screen. But then there is the issue of where to put auto-attack. Maybe spacebar?

**Comment by waveyl** [comment]
Just a reminder: It's difficult for Android to input F1-F12 and Ctrl if the application dosen't provide a convenient method.

**Reply by Zireael07** [reply]
I fully expect Android to have a separate keymap (like it kinda already does ;) )

**Reply by I-am-Erk** [reply]
Just so, Android and other unique setups will need their own maps and are mostly irrelevant to this part.

**Reply by Dansiman** [reply]
FWIW, this is easily addressed with the [Hacker's Keyboard](https://play.google.com/store/apps/details?id=org.pocketworkstation.pckeyboard) app.

**Comment by Altogolik** [comment]
I hope y, u, h, j, k, l, b, n movement keys will be removed from the standard layout

**Comment by Astra-rf** [comment]
I believe that the individual 'look' keys should be removed in favour of the 'Look Around' key which fulfils the same job.

**Comment by brokeboiflex** [comment]
Horizontal scrolling should be standardized across different menus as it's the most difficult keybind to learn

**Comment by Dansiman** [comment]
This might be beyond the scope of this discussion, but Nethack has a set of "extended commands" for the less-frequently-used commands, so that they don't have to occupy a binding. After pressing `#`, you get a prompt to type in a word and press `Enter` to execute one of those rarely-used commands. (Example: `#twoweapon` to toggle two-weapon fighting mode; some never use it, others might turn it on once and then never turn it off.)
CDDA has a lot more commands than Nethack, so perhaps a similar system could be beneficial. Also worth noting that most of the commands that _do_ have their own keybinds by default _also_ have a keyword that can be used with `#`, so that players who wish to bind some of those commands that by default are only accessible via `#` can easily access the command that they unbind via that same system.
While any implementation of this would be its own project, it might yet be worth considering how that feature might relieve any new set of default bindings from being obligated to include a given command merely because it can't (currently) be accessed at all if not bound to a specific key.
